distbench_threadpool.cc

```diff
@@ -42,40 +67,114 @@ DistbenchThreadpool::DistbenchThreadpool(int nb_threads) {
+CThreadpool::CThreadpool(int nb_threads) { thpool_ = thpool_init(nb_threads); }
+
+CThreadpool::~CThreadpool() {
   thpool_wait(thpool_);
   thpool_destroy(thpool_);
 }
 
-void Trampoline(void* function) {
-  auto f = reinterpret_cast<std::function<void()>*>(function);
+void CThreadpool::AddWork(std::function<void()> function) {
+  // Copy the functor object to the heap, and pass the address of the heap
+  // object to the thread pool:
+  auto* fpointer = new std::function<void()>(function);
+  thpool_add_work(thpool_, Trampoline, fpointer);
+}
+
+void CThreadpool::Trampoline(void* heap_object_pointer) {
+  // Execute and delete the heap allocated copy of the functor object:
+  auto f = reinterpret_cast<std::function<void()>*>(heap_object_pointer);
   (*f)();
   delete f;
 }
 
-void DistbenchThreadpool::AddWork(std::function<void()> function) {
-  auto* fpointer = new std::function<void()>(function);
-  thpool_add_work(thpool_, Trampoline, fpointer);
+#ifdef WITH_MERCURY
+class MercuryThreadpool : public AbstractThreadpool {
+ public:
+  MercuryThreadpool(int nb_threads);
+  virtual ~MercuryThreadpool();
+  virtual void AddWork(std::function<void()> function);
+
+ private:
+  struct HeapObject {
+    struct hg_thread_work work_item; /* Must be first! */
+    std::function<void()> function;
+  };
+  static void* Trampoline(void* heap_object_pointer);
+  std::unique_ptr<hg_thread_pool_t> thread_pool_;
+};
+
+MercuryThreadpool::MercuryThreadpool(int nb_threads) {
+  hg_thread_pool_t* tmp;
+  int error = hg_thread_pool_init(nb_threads, &tmp);
+  thread_pool_.reset(tmp);
+  if (error) {
+    LOG(INFO) << "some kinda error: " << error;
+  }
+}
+
+MercuryThreadpool::~MercuryThreadpool() {
+  hg_thread_pool_destroy(thread_pool_.release());
 }
 
+void MercuryThreadpool::AddWork(std::function<void()> function) {
+  // Copy the functor object to the heap, and pass the address of the heap
+  // object to the thread pool:
+  auto* heap_object = new HeapObject;
+  heap_object->work_item.func = Trampoline;
+  heap_object->work_item.args = heap_object;
+  heap_object->function = function;
+  hg_thread_pool_post(thread_pool_.get(), &heap_object->work_item);
+}
+
+void* MercuryThreadpool::Trampoline(void* heap_object_pointer) {
+  // Execute and delete the heap allocated copy of the functor object:
+  auto heap_object = reinterpret_cast<HeapObject*>(heap_object_pointer);
+  heap_object->function();
+  delete heap_object;
+  return 0;
+}
+#endif  // WITH_MERCURY
+}  // anonymous namespace
+
+std::unique_ptr<AbstractThreadpool> CreateThreadpool(std::string_view name,
+                                                     int size) {
+  if (name.empty() || name == "simple") {
+    return std::make_unique<SimpleThreadpool>(size);
+  } else if (name == "cthread") {
+    return std::make_unique<CThreadpool>(size);
+#ifdef WITH_MERCURY
+  } else if (name == "mercury") {
+    return std::make_unique<MercuryThreadpool>(size);
 #endif
+  } else {
+    return {};
+  }
+}
```

Comment:
Also for my understanding, is it possible to add a one line comment at the top of this file that explains when developers should use a simple thread Vs cthread vs mercuryThread?

Reply:
the selection comes from the user via the test configuration, so this belongs in user documentation.